### PR TITLE
Memory and stability fixes to XLForm

### PR DESCRIPTION
**Pull request by liraz**
The following fixes helped our app to be a lot more stable in production with thousands of users using the app heavily.
library before the fixes had many memory issues.
please review the fixes - they should be a great benefit to you.

### CODE REVIEW

File: XLForm/XL/Helpers/NSObject+XLFormAdditions.m [hunk 1]
```diff
@@ -53,7 +53,7 @@ -(id)valueData
         return result;
     }
     if ([self conformsToProtocol:@protocol(XLFormOptionObject)]){
-        return [(id<XLFormOptionObject>)self formValue];
```
Comment (mats-claassen) [inline XLForm/XL/Helpers/NSObject+XLFormAdditions.m]: Why is this change?

File: XLForm/XL/Helpers/UIView+XLFormAdditions.m [hunk 2]
```diff
@@ -27,47 +27,60 @@
...
+                                       multiplier:1.0
+                                         constant:0.0];
 }
 
 - (UIView *)findFirstResponder
 {
+    UIView *firstResponder = nil;
     if (self.isFirstResponder) {
-        return self;
+        firstResponder = self;
```
Comment (mats-claassen) [inline XLForm/XL/Helpers/UIView+XLFormAdditions.m]: I do not understand this change. If the view is the first responder then the search finished and `self` should be returned, right?

File: XLForm/XL/Helpers/UIView+XLFormAdditions.m [hunk 3]
```diff
@@ -27,47 +27,60 @@
...
     }
     for (UIView *subView in self.subviews) {
-        UIView *firstResponder = [subView findFirstResponder];
-        if (firstResponder != nil) {
-            return firstResponder;
+        UIView *fr = [subView findFirstResponder];
```
Comment (mats-claassen) [inline XLForm/XL/Helpers/UIView+XLFormAdditions.m]: Why di you change the name of the variable. `firstResponder` is better to understand here

File: XLForm/XL/Cell/XLFormTextViewCell.m [hunk 4]
```diff
@@ -70,34 +70,26 @@ -(void)dealloc
...
-
 #pragma mark - XLFormDescriptorCell
 
 -(void)configure
 {
     [super configure];
     [self setSelectionStyle:UITableViewCellSelectionStyleNone];
-    [self.contentView addSubview:self.textLabel];
-    [self.contentView addSubview:self.textView];
+    UILabel *textLabel = [UILabel autolayoutView];
```
[the remark this answers is not in the document]
Comment (iKorich) [reply XLForm/XL/Cell/XLFormTextViewCell.m]: to remove extra check

File: XLForm/XL/Helpers/UIView+XLFormAdditions.m (hunk 2 above)
Comment (iKorich) [reply XLForm/XL/Helpers/UIView+XLFormAdditions.m]: to make one exit from function

File: XLForm/XL/Helpers/NSObject+XLFormAdditions.m (hunk 1 above)
Comment (iKorich) [reply XLForm/XL/Helpers/NSObject+XLFormAdditions.m]: reverted

File: XLForm/XL/Cell/XLFormSelectorCell.m [hunk 5]
```diff
@@ -57,19 +57,33 @@ -(NSString *)valueDisplayText
...
+                    }
+                    else{
+                        [descriptionArray addObject:[option displayText]];
                     }
                 }
-                else{
-                    [descriptionArray addObject:[option displayText]];
+            }
+        }else{
+            for (id option in self.rowDescriptor.value) {
```
[the remark this answers is not in the document]
Comment (iKorich) [reply XLForm/XL/Cell/XLFormSelectorCell.m]: I got that from issue #948, can't remember why, now it doesn't use for us, reverted

File: XLForm/XL/Helpers/UIView+XLFormAdditions.m (hunk 2 above)
Comment (mats-claassen) [reply XLForm/XL/Helpers/UIView+XLFormAdditions.m]: Okay, but we should not loop through the `subviews` if `self.isFirstResponder`

Comment (iKorich) [reply XLForm/XL/Helpers/UIView+XLFormAdditions.m]: fixed in 509ea4d

File: XLForm/XL/Descriptors/XLFormRowDescriptor.m [hunk 6]
```diff
@@ -132,7 +143,8 @@ -(XLFormBaseCell *)cellForFormController:(XLFormViewController * __unused)formCo
                 NSString *folderName = [components firstObject];
                 NSString *bundlePath = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:folderName];
                 bundle = [NSBundle bundleWithPath:bundlePath];
-            } else {
```
Comment (mats-claassen) [inline XLForm/XL/Descriptors/XLFormRowDescriptor.m]: This change does not follow the ObjC guidelines. For example [here](https://github.com/raywenderlich/objective-c-style-guide#spacing)

Comment (mats-claassen) [reply XLForm/XL/Descriptors/XLFormRowDescriptor.m]: Same happens several times below

File: XLForm/XL/Descriptors/XLFormRowDescriptor.m [hunk 7]
```diff
@@ -260,52 +290,46 @@ -(void)setHeight:(CGFloat)height {
...
-    rowDescriptorCopy->_hidden = _hidden;
-    rowDescriptorCopy->_disabled = _disabled;
+    rowDescriptorCopy.hidden = self.hidden;
+    rowDescriptorCopy.disabled = self.disabled;
     rowDescriptorCopy.required = self.isRequired;
     rowDescriptorCopy.isDirtyDisablePredicateCache = YES;
     rowDescriptorCopy.isDirtyHidePredicateCache = YES;
     rowDescriptorCopy.validators = [self.validators mutableCopy];
-
+
```
Comment (mats-claassen) [inline XLForm/XL/Descriptors/XLFormRowDescriptor.m]: Again. remove trailing whitespaces. This goes against style guidelines

**Comment by mats-claassen** [comment]
@liraz @iKorich Please open new PRs for further changes. Thanks for contributing

**Comment by iKorich** [comment]
1 month of tests - there are no crashes on memory warnings

**Pull request merged**